### PR TITLE
Hardcode the labels stable hash function instead of taking it as an option

### PR DESCRIPTION
**Pull request by pracucci**
This PR is a follow up of https://github.com/grafana/mimir-prometheus/pull/369. Instead of taking the sharding function as an option, we always use a "stable" hashing function which is guaranteed to not change over time.

### CODE REVIEW

File: tsdb/head.go [hunk 1]
```diff
@@ -1568,12 +1566,8 @@ func (h *Head) getOrCreate(hash uint64, lset labels.Labels) (*memSeries, bool, e
 }
 
 func (h *Head) getOrCreateWithID(id chunks.HeadSeriesRef, hash uint64, lset labels.Labels) (*memSeries, bool, error) {
-	shardHash := hash
-	if h.opts.ShardFunc != nil {
-		shardHash = h.opts.ShardFunc(lset)
-	}
 	s, created, err := h.series.getOrSet(hash, lset, func() *memSeries {
-		return newMemSeries(lset, id, shardHash, h.opts.ChunkEndTimeVariance, h.opts.IsolationDisabled)
+		return newMemSeries(lset, id, labels.StableHash(lset), h.opts.ChunkEndTimeVariance, h.opts.IsolationDisabled)
```
Comment (pracucci) [inline tsdb/head.go:1570]: Note to reviewers: this callback is not called within a lock, so should be fine moving the hashing here.

File: tsdb/head.go [hunk 2]
```diff
@@ -1792,7 +1786,7 @@ func (s *stripeSeries) getOrSet(hash uint64, lset labels.Labels, createSeries fu
 	// rather it gives a 'hint' whether to create a series or not.
 	preCreationErr := s.seriesLifecycleCallback.PreCreation(lset)
 
-	// Create the series, unless the PreCreation() callback as failed.
+	// Create the series, unless the PreCreation() callback has failed.
```
Comment (bboreham) [inline tsdb/head.go:1789]: Best to do this in a separate PR upstream then merge to here.

**Comment by pracucci** [comment]
> As mentioned in Mimir's PR, we'd need to make sure that there's a test ensuring that we're using this specific hashing algorightm.

Yes, I'm wondering on it, but it's a set of tests we'll have in Mimir.

File: model/labels/sharding.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+// StableHash is a labels hashing implementation which is guaranteed to not change over time.
+// This function should be used whenever labels hashing backward compatibility must be guaranteed.
+func StableHash(l Labels) uint64 {
+	hash, _ := labelsXXHash(make([]byte, 0, 1024), l)
+	return hash
+}
+
+// labelsXXHash replicates historical hash computation from Prometheus circa 2022.
+// The buffer b can be recycled to avoid allocations.
+func labelsXXHash(b []byte, ls Labels) (uint64, []byte) {
```
Comment (bboreham) [inline model/labels/sharding.go:16]: I think having a separate function was only useful if it could be called with a reused buffer, and this ability is not used in this PR.
Maybe profiling will tell us we need that, but otherwise you can elide the two functions.

**Review by bboreham — COMMENTED** [review]
lgtm; couple of thoughts below

File: tsdb/head.go (hunk 2 above)
Comment (pracucci) [reply tsdb/head.go]: Yep, done.

File: model/labels/sharding.go (hunk 3 above)
Comment (pracucci) [reply model/labels/sharding.go]: Yep, done.

**Comment by krajorama** [comment]
late to the party, but +1
